packages/webpack/css-extract-webpack-plugin/src/CssExtractRspackPlugin.ts

```diff
@@ -169,6 +147,109 @@ class CssExtractRspackPluginImpl {
+          if (cssHmrJSONGenerated) return;
+          if (!baseEncodeOptions || !assets) return;
+          // Generate the CSS HMR JSON for each CSS file
+          for (const [filename, source] of Object.entries(assets)) {
+            if (!filename.endsWith('.css')) {
+              continue;
+            }
+            const content: string = source.source().toString('utf-8');
+            const { cssMap } = LynxTemplatePlugin.convertCSSChunksToMap(
+              [content],
+              options.cssPlugins,
+              Boolean(baseEncodeOptions.compilerOptions['enableCSSSelector']),
+            );
+            const cssDeps = Object.entries(cssMap).reduce<
+              Record<string, string[]>
+            >((acc, [key, value]) => {
+              const importRuleNodes = value.filter(
+                (node) => node.type === 'ImportRule',
+              );
+
+              acc[key] = importRuleNodes.map(({ href }) => href);
+              return acc;
+            }, {});
+
+            try {
+              const encoded = await LynxEncodePlugin.encodeCSS(
+                [content],
+                baseEncodeOptions,
+                options.cssPlugins,
+                hooks.encode.taps.length > 1
+                  ? async (encodeOptions: EncodeOptions) => {
+                    return await hooks.encode.promise({
+                      encodeOptions,
+                      templateType: 'css-hmr',
+                    });
+                  }
+                  : undefined,
+              );
+              const result = {
+                content: encoded.toString('base64'),
+                deps: cssDeps,
+              };
+              compilation.emitAsset(
+                filename.replace(
+                  '.css',
+                  `${this.hash ? `.${this.hash}` : ''}.css.hot-update.json`,
+                ),
+                new compiler.webpack.sources.RawSource(
+                  JSON.stringify(result),
+                  true,
+                ),
+              );
+            } catch (error) {
+              if (
+                error && typeof error === 'object' && 'error_msg' in error
+              ) {
+                compilation.errors.push(
+                  // TODO: use more human-readable error message(i.e.: using sourcemap to get source code)
+                  //       or give webpack/rspack with location of bundle
+                  new compiler.webpack.WebpackError(
+                    error.error_msg as string,
+                  ),
+                );
+              } else {
+                compilation.errors.push(
+                  error as (typeof compilation.errors)[0],
+                );
+              }
+            }
+          }
+          this.hash = compilation.hash;
+          cssHmrJSONGenerated = true;
+        };
+
+        // @ts-expect-error only tap to get the encode options, not existing early
+        hooks.encode.tapPromise(this.name, async (args) => {
```

Comment:
I don't think we should get the encode options here. You may use the `hooks.beforeEmit`, which has the final encode options.

Reply:
@colinaaa I considered choosing `encode` because other plugins might modify the encode option by the `hooks.beforeEmit` hook, in order to get the final encode option used by the template assembler, I used `hooks.encode`.

Maybe I should change to `hooks.beforeEmit` with a large `stage` number to ensure our tap is executed in the last order, what do you think?